terraform/ec2-examples/distributed-ml-training-fsdp/README.md

```diff
@@ -0,0 +1,143 @@
+
+![Distributed ML architecture](../../../docs/distributed-ml-training-architecture.png)
+
+By default, this blueprint uses a set of three g5.12xlarge (with 4 GPUs) instances to showcase a multi-node, fully sharded data parallel distributed training. You can modify this blueprint to use larger instances from the local variable **instance_type_workers** and **instance_type_head** if you need more GPUs. - if you change the instance type, you need to also modify the worker task and service definition memory, CPU and GPUs and the container command parameters. The [training script example](./training_example.py) assumes 2 machines with a single GPU each, but can be changed via the **num_workers** variable.
+
+## Components
+
+* Service discovery: The head node is registered to a private DNS using local zones via cloud map. This allows worker tasks to discover the head task and join the cluster on start-up.
+* 2 autoscaling groups: One for the head instance and another one for the worker instances
+* ECS service definition:
+    * Head service: runs singleton processes responsible for cluster management along with training jobs
+    * Worker service: runs training jobs
+* S3 bucket to store as shared storage
+
+## Deployment
+
+1. Deploy core-infra resources
+
+```shell
+cd ./terraform/ec2-examples/core-infra
+terraform init
+terraform apply -target=module.vpc -target=aws_service_discovery_private_dns_namespace.this
+```
+
+2. Deploy this blueprint
+
+```shell
+cd ../distributed-ml-training-fsdp
+terraform init
+terraform apply
+```
+
+It can take several minutes until the instances are created and connected to SSM
+
+## Example: training the databricks/dolly-v2-7b model with the tiny_shakespeare dataset dataset
+
+Once the instances are running, you can connect to the EC2 instance running the head container using SSM, and open a bash shell in the container from there. This is only for demonstration purposes - Using notebooks with [SageMaker](https://aws.amazon.com/sagemaker/) or [Cloud 9](https://aws.amazon.com/cloud9/) provide a better user experience to run training jobs in python than using the bash shell.
+
+1. Connect to the head instance
+
+```bash
+HEAD_INSTANCE_ID=$(aws ec2 describe-instances \
+  --filters 'Name=tag:Name,Values=ecs-demo-distributed-ml-training-head' 'Name=instance-state-name,Values=running' \
+  --query 'Reservations[*].Instances[0].InstanceId' --output text --region us-west-2
+)
+aws ssm start-session --target $HEAD_INSTANCE_ID --region us-west-2
+```
+
+2. Connect to the container
+
+Due to the size of the container images, it might take several minutes until the containers reach a running state. The following command will fail if the container is not running.
+
+```
+CONTAINER_ID=$(sudo docker ps -qf "name=.*-rayhead-.*")
+sudo docker exec -it $CONTAINER_ID bash
+```
+
```

Comment:
Can’t you use ecs exec instead to con ext to the container ?

Reply:
I would prefer to not because it will add more complexity to the deployment and it will only save 2 commands. I will add a note that it is available if needed.

Reply:
No problem, I enabled ECS exec in the head task and updated the README